### PR TITLE
Fixed #31331 -- Switched MySQL to group by selected primary keys. 

### PR DESCRIPTION
**Pull request by charettes**
ticket-31331

### CODE REVIEW

File: tests/annotations/tests.py [hunk 1]
```diff
@@ -550,21 +550,6 @@ def test_values_with_pk_annotation(self):
...
-            .annotate(
-                min_book_id=raw,
-                count_friends=Count("friends"),
-            )
-            .order_by()
-        )
-        _, _, group_by = qs.query.get_compiler(using="default").pre_sql_setup()
-        self.assertEqual(len(group_by), 1)
-        self.assertNotEqual(raw, group_by[0])
-
```
Comment (charettes) [inline tests/annotations/tests.py:567]: This can now easily be achieved by using a `Subquery` which will do the right thing.

**Comment by charettes** [comment]
Well that's an interesting one, MySQL always require columns in `HAVING` to be present in the `GROUP BY` independently of whether or not they are functionally dependent otherwise it errors out with an unhelpful `"Unknown column 'col' in 'having clause'"`

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -235,7 +204,9 @@ def collapse_group_by(self, expressions, having):
             expressions = [
                 expr
                 for expr in expressions
-                if expr in pks or getattr(expr, "alias", None) not in aliases
+                if expr in pks
+                or expr in having
```
Comment (charettes) [inline django/db/models/sql/compiler.py:208]: Even if not required on Postgres I figured the simplicity tradeoff this brings to `collapse_group_by` was worth it.

File: docs/releases/4.2.txt [hunk 3]
```diff
@@ -309,7 +309,12 @@ Database backend API
 This section describes changes that may be needed in third-party database
 backends.
 
-* ...
+* Support for the ``allows_group_by_pks`` feature flag has been removed as it
```
Comment (timgraham) [inline docs/releases/4.2.txt:312]:
```
``DatabaseFeatures.allows_group_by_pks`` is removed as it ...
```

File: docs/releases/4.2.txt [hunk 4]
```diff
@@ -309,7 +309,12 @@ Database backend API
...
+  only remained to accommodate a MySQL extension that has been supplanted by proper
+  functional dependency detection in MySQL 5.7.15. Note that the
+  ``allows_group_by_selected_pks`` feature flag is still supported and should
```
Comment (timgraham) [inline docs/releases/4.2.txt:315]:
```
``DatabaseFeatures.allows_group_by_selected_pks`` is still supported...
```

**Comment by charettes** [comment]
Thanks for the wording review as usual @timgraham.

File: docs/releases/4.2.txt [hunk 5]
```diff
@@ -309,7 +309,12 @@ Database backend API
 This section describes changes that may be needed in third-party database
 backends.
 
-* ...
+* ``DatabaseFeatures.allows_group_by_pks`` is removed as it only remained to
```
Comment (felixxm) [inline docs/releases/4.2.txt:312]:
```suggestion
* ``DatabaseFeatures.allows_group_by_pk`` is removed as it only remained to
```

**Comment by felixxm** [comment]
@charettes Thanks :star: I pushed tiny edits.

**Pull request merged**